contracts/adapters/OptimisticAuctionRebalanceExtensionV1.sol

```diff
@@ -0,0 +1,406 @@
+     * @param _oldComponents                Addresses of existing components in the SetToken.
+     * @param _newComponents                Addresses of new components to be added.
+     * @param _newComponentsAuctionParams   AuctionExecutionParams for new components, indexed corresponding to _newComponents.
+     * @param _oldComponentsAuctionParams   AuctionExecutionParams for existing components, indexed corresponding to
+     *                                      the current component positions. Set to 0 for components being removed.
+     * @param _rebalanceDuration            Duration of the rebalance in seconds.
+     * @param _positionMultiplier           Position multiplier at the time target units were calculated.
+    */
+    function proposeRebalance(
+        IERC20 _quoteAsset,
+        address[] memory _oldComponents,
+        address[] memory _newComponents,
+        AuctionExecutionParams[] memory _newComponentsAuctionParams,
+        AuctionExecutionParams[] memory _oldComponentsAuctionParams,
+        uint256 _rebalanceDuration,
+        uint256 _positionMultiplier
+    )
+        external
+        onlyAllowedAssets(_newComponents)
+        onlyIfOpen()
+    {
+        bytes32 proposalHash = keccak256(abi.encode(
+            setToken,
+            _quoteAsset,
+            _oldComponents,
+            _newComponents,
+            _newComponentsAuctionParams,
+            _oldComponentsAuctionParams,
+            false, // We don't allow locking the set token in this version
+            _rebalanceDuration,
+            _positionMultiplier
+        ));
+        require(assertionIds[proposalHash] == bytes32(0), "Proposal already exists");
+        require(productSettings.rulesHash != bytes32(""), "Rules not set");
+        require(address(productSettings.optimisticParams.optimisticOracleV3) != address(0), "Oracle not set");
+
+
+        bytes memory claim = _constructClaim(proposalHash, productSettings.rulesHash);
+        uint256 totalBond = _pullBond(productSettings.optimisticParams);
+
+        bytes32 assertionId = productSettings.optimisticParams.optimisticOracleV3.assertTruth(
+            claim,
+            msg.sender,
+            address(this),
+            address(0),
+            productSettings.optimisticParams.liveness,
+            productSettings.optimisticParams.collateral,
+            totalBond,
+            productSettings.optimisticParams.identifier,
+            bytes32(0)
+        );
+
+        assertionIds[proposalHash] = assertionId;
+        proposedProduct[assertionId] = Proposal({
+            proposalHash: proposalHash,
+            product: setToken
+        });
+
+        emit RebalanceProposed( setToken, _quoteAsset, _oldComponents, _newComponents, _newComponentsAuctionParams, _oldComponentsAuctionParams,  _rebalanceDuration, _positionMultiplier);
+        emit AssertedClaim(setToken, msg.sender, productSettings.rulesHash, assertionId, claim);
```

Comment:
Let's add unit tests for these event emissions

Reply:
Done: https://github.com/IndexCoop/index-coop-smart-contracts/commit/4d6113f9a42203ee9e3f55ac8a6a93feb37302a4